[CDAP-7624] Fix issue where readless incrememts from different MR tasks cancel out each other

anew opened the pull request:
Issue is that all tasks write a delta with the same version. They don't check whether there is already a delta with the same write version. The fix is this:
- introduce a new runtime argument (to be used in MR) "safe.readless.increments"
- if this argument is true, then HBaseTable, upon persist(), issues an increment() to HBase instead of a put()
- the Increment() is handled region-server-side in a new coprocessor hook. It performs a get of the counters to be incremented, adds the increments, and then puts them back. All this happens under row lock. 
- this needs to be copied to all hbase-compat modules (lots of duplication!) 

Open questions: 
- How can we test this for all HBase versions? It turns out that we do not have any tests for readless increments with transactions in the compat modules. That means right now, they are only tested for HBase-0.98 (which is used by HBaseTableTest)
- HBase 0.96 does not have a coprocessor hook for increments while holding the lock. It is not clear how race conditions can be avoided there. Is it worth doing a different design for 0.96?
- Should we rely on the user to set this runtime argument when a table is used for readless increments in MR? Or should the MR dataset context always add this argument to all datasets?

When reviewing, please only review one compat module. The changes for other compat modules are identical (except for 0.96, it uses the preIncrement() instead of preIncrementAfterRowLock().

Commit:
[CDAP-7624] Fix issue where readless incrememts from different MR tasks cancel out each other

anew commented:
Decision made: The MapReduceTaskContext will always inject the runtime argument for safe readless increments. 

For testing against all HBase versions: Added an integration test here: https://github.com/caskdata/cdap-integration-tests/pull/556. Please review also, it is small.

anew commented:
By the way, no documentation changes, the developer does not need to know about this new runtime argument.

chtyim commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseTable.java:
```diff
@@ -187,28 +193,42 @@ protected void persist(NavigableMap<byte[], NavigableMap<byte[], Update>> update
     }
 
     List<Put> puts = Lists.newArrayList();
+    List<Increment> increments = Lists.newArrayList();
```
I believe you can combine these two lists to one `List<Mutation>` and use `HTable.batch` to execute all puts and increments in one batch.

anew replied:
ok, I can try that

chtyim commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseTable.java:
```diff
@@ -243,6 +276,19 @@ private Put getIncrementalPut(Put existing, byte[] row) {
 
+  private IncrementBuilder getIncrement(IncrementBuilder existing, byte[] row, boolean transactional)
+    throws IOException {
+    if (existing != null) {
```
I don't think the builder is reusable as it never cleans the columns in the internal HBase `Increment` object. The same applies for the existing method that reuses the put builder.

anew replied:
wow. That means we have had a bug in the existing code... so this needs to be cleared at the start of a transaction.

chtyim replied:
That's the sad truth. We should have test case that having multiple rows updating the same column, committing in the same tx.

anew replied:
I think we are good, actually. Because the IncrementBuilder and the PutBuilder are both reset to null every time we start to persist another row. https://github.com/caskdata/cdap/pull/7308/files#diff-3940118a7982e1cceeaa023c99b1b35fR200

chtyim commented on cdap-hbase-compat-0.96/src/main/java/co/cask/cdap/data2/increment/hbase96/IncrementHandler.java:
```diff
@@ -120,6 +124,67 @@ public void preGetOp(ObserverContext<RegionCoprocessorEnvironment> ctx, Get get,
+                                           + " must be set for transactional readless increments");
+    }
+    long tsToAssign;
+    if (transactional) {
```
If it is not transactional, isn't that we can just turn the `Increment` into a `Put` and call `region.Put` (with bypass as well) without doing the rest of the logic? The `Put` should trigger the `prePut` logic and do the right thing?

anew replied:
yeah probably. In the non-tx case, HBase will assign a unique ts.

chtyim commented on cdap-hbase-compat-0.96/src/main/java/co/cask/cdap/data2/increment/hbase96/IncrementHandler.java:
```diff
@@ -120,6 +124,67 @@ public void preGetOp(ObserverContext<RegionCoprocessorEnvironment> ctx, Get get,
+      tsToAssign = state.getUniqueTimestamp();
+    }
+
+    Get get = new Get(increment.getRow());
```
Shall we acquire the row lock to cover the get and put operations? Something like:

```java
region.getRowLock(increment.getRow());
try {
  // wait for all prior MVCC transactions to finish - while we hold the row lock
  // (so that we are guaranteed to see the latest state). Copied from Region server code
  MultiVersionConsistencyControl mvcc = region.getMVCC();
  mvcc.completeMemstoreInsert(mvcc.beginMemstoreInsert());
  
  Get get = new Get(....);
  ...
} finally {
  rowLock.release();
}
```

anew replied:
I can try that. I am not that familiar with the HBase internal MVCC APIs... but I can follow your suggestion

chtyim replied:
I am also just reading it off from the Region server code. I remember I used it once on the Flow queue coprocessor (but that logic is no longer there since now we don't do Get in that coprocessor any more)

anew commented:
@chtyim I addressed your comments, please take another look.

Note the changes to all the IncrementHandlers are the same. Except that the .96 has an extra change to acquire the row lock.

chtyim commented:
LGTM

anew commented:
waiting for build to pass. http://builds.cask.co/browse/CDAP-DUT5183

anew commented:
build passed. squashing and merging.
